nbft.c

```diff
@@ -0,0 +1,691 @@
+		flags = NORMAL;
+	else if (!strcmp(format, "json"))
+		flags = JSON;
+	else
+		return EINVAL;
+
+	if (!strcmp(config_file, "none"))
+		config_file = NULL;
+
+	r = nvme_create_root(stderr, map_log_level(verbose, false));
+	if (!r) {
+		fprintf(stderr, "Failed to create topology root: %s\n",
+			nvme_strerror(errno));
+		return -errno;
+	}
+
+	ret = nvme_scan_topology(r, NULL, NULL);
+	if (ret < 0) {
+		fprintf(stderr, "Failed to scan topology: %s\n",
+			nvme_strerror(errno));
+		nvme_free_tree(r);
+	}
+	nvme_read_config(r, config_file);
+
+	if (host_traddr)
+		user_host_traddr = true;
+	if (hostnqn)
+		user_hostnqn = true;
+	if (hostid)
+		user_hostid = true;
+
+	list_head_init(&nbft_list);
+	ret = read_sysfs_nbft_files(&nbft_list, nbft_path);
+	if (ret)
+		goto out_free_2;
+
+	list_for_each(&nbft_list, nbft, node)
+		list_for_each(&nbft->subsystem_ns_list, ss, node)
+			for (i = 0; i < ss->num_hfis; i++) {
+				hfi = ss->hfis[i];
+				free_hnqn = false;
+				if (!user_hostnqn) {
+					hostnqn = hnqn = nbft->host.nqn;
+					if (!hostnqn) {
+						hostnqn = hnqn = nvmf_hostnqn_from_file();
+						free_hnqn = true;
+					}
+				}
+
+				free_hid = false;
+				if (!user_hostid) {
+					if (*nbft->host.id) {
+						hostid = hid = (char *)util_uuid_to_string(*nbft->host.id);
+						if (!hostid) {
+							hostid = hid = nvmf_hostid_from_file();
+							free_hid = true;
+						}
+					}
+				}
+
+				h = nvme_lookup_host(r, hostnqn, hostid);
+				if (!h) {
+					errno = ENOMEM;
+					goto out_free;
+				}
+
+				if (!user_host_traddr) {
+					host_traddr = NULL;
+					if (!strncmp(ss->transport, "tcp", 3))
+						host_traddr = hfi->tcp_info.ipaddr;
+				}
+
+				//if (hostkey)
+				//	nvme_host_set_dhchap_key(h, hostkey);
+				c = nvme_create_ctrl(r, ss->subsys_nqn, ss->transport, ss->transport_address,
+						     host_traddr, NULL, ss->transport_svcid);
+				if (!c) {
+					errno = ENOMEM;
+					goto out_free;
+				}
+				//if (ctrlkey)
+				//	nvme_ctrl_set_dhchap_key(c, ctrlkey);
+
+				errno = 0;
+				ret = nvmf_add_ctrl(h, c, &cfg);
+
+				/*
+				 * With TCP/DHCP, it can happen that the OS
+				 * obtains a different local IP address than the
+				 * firwmare had. Retry without host_traddr.
+				*/
+				if (ret == -1 && errno == ENVME_CONNECT_WRITE &&
+				    host_traddr && !user_host_traddr &&
+				    !strcmp(ss->transport, "tcp") &&
+				    strlen(hfi->tcp_info.dhcp_server_ipaddr) > 0) {
+					nvme_free_ctrl(c);
+					c = nvme_create_ctrl(r, ss->subsys_nqn, ss->transport,
+							     ss->transport_address,
+							     NULL, NULL, ss->transport_svcid);
+					if (!c) {
+						errno = ENOMEM;
+						goto out_free;
+					}
+					errno = 0;
+					ret = nvmf_add_ctrl(h, c, &cfg);
+					if (ret == 0 && verbose >= 1)
+						fprintf(stderr,
+							"connect with host_traddr=\"%s\" failed, success after omitting host_traddr\n",
+							host_traddr);
+				}
+
+				if (ret)
+					fprintf(stderr, "no controller found\n");
```

Comment:
We should handle the case `ENVME_CONNECT_ALREADY` here. Dracut may need to do multiple connection attempts. "no controller found" is not a helpful error message if the subsystem at hand is already connected.

Reply:
There is `const char *nvme_status_to_string(int status, bool fabrics);` which translates the error codes into something more useful.

Reply:
We might actually consider not returning an error if we find that (all?) connections specified in the NBFT are already set up.

Reply:
FWIW `nvme connect-all` ignores already connected targets. It only logs the warning.

Reply:
@mwilck are we good here?  Do you still want to make this change?